[Security Solution] Use CellActions registry in Discover data grid

semd opened the pull request:
## Summary

closes: https://github.com/elastic/kibana/issues/157191

Enables Discover DataGrid to use registered cell actions instead of the default static actions.

### New `cellActionsTriggerId` prop
This PR introduces a new `cellActionsTriggerId` _optional_ prop in the DataGrid component:

https://github.com/elastic/kibana/blob/98c210f9ecce969d3621daa3b009aced80ef8f92/src/plugins/discover/public/components/discover_grid/discover_grid.tsx#L198-L201

When this prop is defined, the component queries the trigger's registry to retrieve the cellActions attached to it, using the CellActions package' `useDataGridColumnsCellActions` hook. This hook returns the cellActions array ready to be passed for each column to the EuiDataGrid component.
When (non-empty) actions are found in the registry, they are used, replacing all of the default static Discover actions. Otherwise, the default cell actions are used.

This new prop also allows other instances of the Discover DataGrid to be configured with custom cell actions, which will probably be needed by Security Timeline integration with Discover.

### New `SEARCH_EMBEDDABLE_CELL_ACTIONS_TRIGGER` Trigger

Along with the new `cellActionsTriggerId` prop the plugin also registers a new trigger for "saved search" embeddable:

https://github.com/elastic/kibana/blob/055750c8ddf6859d89c1dd89ab22bd03273a2d82/src/plugins/discover/public/plugin.tsx#L387

And it gets passed to the DataGrid component on the Embeddable creation:

https://github.com/elastic/kibana/blob/055750c8ddf6859d89c1dd89ab22bd03273a2d82/src/plugins/discover/public/embeddable/saved_search_embeddable.tsx#L403

Having this new trigger available allows solutions to attach custom actions to it, in order to be displayed in the saved search embeddables. Each action will be able to implement its `isCompatible` check to determine if they are going to be displayed in the embedded saved search DataGrid field, or not. If no compatible actions are found, DataGrid will render the default static actions.


ℹ️  In this implementation, the actions registered to this new "embeddable trigger" need to check if they are being rendered inside Security using the `isCompatible` function, to prevent them from being displayed in other solutions, resulting in a non-optimal architecture. This approach was needed since there's no plausible way to pass the `cellActionsTriggerId` property from the Dashboard Renderer used in Security, all the way down to the specific Discover "saved search" embeddable. However, the Dashboards team is planning to enable us to pass options to nested embeddables using a registry (https://github.com/elastic/kibana/issues/148933). When this new tool is available we will be able to delegate the trigger registering to Security and configure the "saved search" embeddables to use it. Therefore, the trigger will only be used by Security, so we won't have to worry about Security actions being rendered outside Security.


## Videos

before:

https://github.com/elastic/kibana/assets/17747913/de92cd74-6125-4766-8e9d-7e0985932618

after:

https://github.com/elastic/kibana/assets/17747913/f9bd597a-860e-4572-aa9d-9f1c72c11a4b

semd committed:
initial working version

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

semd committed:
tests

semd committed:
Merge branch 'poc_discover_custom_cell_actions' of https://github.com/semd/kibana into poc_discover_custom_cell_actions

semd committed:
conflicts solved

semd committed:
new test cases

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

semd commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into poc_discover_custom_cell_actions

semd committed:
Merge remote-tracking branch 'upstream/main' into poc_discover_custom_cell_actions

semd committed:
fix weird dependencies problem

semd committed:
solve conflicts

kertal reviewed (commented):
I think extending the cell actions of Discover data grid makes much sense 👍  
We will have a closer look at it next week. Playing with it, I have an initial question. When I add something to a timeline, what does it mean in this cases? I've created a custom dashboard showing e-commerce orders, I've added some of the values to a timeline. Now those values are visible in the query builder, but that's it. Since the data view Alerts is selected, there are no entries visible in the table. I guess it's expected. But as a newbie user more familiar with Discover / Dashboards it's not 100% clear what add to "Add to timeline" means in this case

semd committed:
Merge remote-tracking branch 'upstream/main' into poc_discover_custom_cell_actions

jughosta reviewed (commented):
Looking good! Left some questions to clarify the details.

semd commented on src/plugins/discover/public/application/context/context_app_content.test.tsx:
```diff
@@ -15,12 +15,16 @@ import { SortDirection } from '@kbn/data-plugin/public';
+import { ServicesContextProvider } from '../services_provider';
 import { buildDataTableRecord } from '../../utils/build_data_record';
 
+const MockDiscoverGrid = () => <></>;
+jest.mock('../../components/discover_grid/discover_grid', () => ({
+  DiscoverGrid: jest.fn(() => <MockDiscoverGrid />),
+}));
```
No, the test was passing, but complaining about an update: `Warning: An update to DiscoverGrid inside a test was not wrapped in act(...).`.
I can add the async `act` after the `mountWithIntl` if you want, but it would meach changing the whole test to async. Since here we were only checking the _DiscoverGrid_ was rendered inside I just mocked the component, it should make no difference.

jughosta replied:
If it's not much of a trouble, would be great to update tests and keep the original `DiscoverGrid`. Thanks for updating other tests!

semd commented on src/plugins/discover/public/components/discover_grid/discover_grid.tsx:
```diff
@@ -416,14 +425,45 @@ export const DiscoverGrid = ({
+          return {
+            name: columnName,
+            type: field?.type ?? 'keyword',
+            values: displayedRows.map((row) => {
+              const value = row.flattened?.[columnName];
+              const valuesArray = Array.isArray(value) ? value : [value];
+              return valuesArray;
+            }),
```
The `useDataGridColumnsCellActions` hook returns the `EuiDataGridColumnCellAction[]` for each field, this is the type that `EuiDataGrid` needs for the column's `cellActions` property, which is an array of functions that only receive `rowIndex` to identify what are the values that we have to use for that specific cell action:

https://github.com/elastic/kibana/blob/76f6fd9ef330c2be34d8151bbccc7b29afe63f92/packages/kbn-cell-actions/src/hooks/use_data_grid_column_cell_actions.tsx#L100

So we need all the row values of each column when we build the whole structure of `EuiDataGridColumnCellAction` functions, in order to extract the proper value from the row index when the action is executed.

Do you think there's any scenario where this could be a problem?

jughosta replied:
Yes, basically the default scenario. Discover should be able to handle large number of rows and columns. Also we are planning to integrate pagination with lazy loading - so number of rows will only grow.

My suggestion would be to update `useDataGridColumnsCellActions` so it can accept a function which would give a value by `rowIndex` and field name in return. Quick example of it:
```
const getCellValue = useCallback((fieldName, rowIndex) => {
 return displayedRows[rowIndex].flattened[fieldName]
}, [displayedRows])

const columnsCellActions = useDataGridColumnsCellActions(cellActionProps, getCellValue);
```

The idea is to avoid unnecessary iterations over fields array or rows array. Could you please double check `fields.map` calls in `useDataGridColumnsCellActions` too? Can it be optimized?

Since EuiGrid supports virtualization and it does not render all data at the same time, we should try to avoid extra calculations over all data too.

semd commented:
> I think extending the cell actions of Discover data grid makes much sense 👍 We will have a closer look at it next week. Playing with it, I have an initial question. When I add something to a timeline, what does it mean in this cases? I've created a custom dashboard showing e-commerce orders, I've added some of the values to a timeline. Now those values are visible in the query builder, but that's it. Since the data view Alerts is selected, there are no entries visible in the table. I guess it's expected. But as a newbie user more familiar with Discover / Dashboards it's not 100% clear what add to "Add to timeline" means in this case

Hi @kertal 
Yes you are right, the scenario you described is the expected behavior. However, these actions will only be visible inside Security, where the dashboards are expected to contain Security-related data. The "add to timeline" action is meant to make it easy for Security analysts to pivot to Timeline investigations on an specific value, probably from the alerts index. 
It does not make much sense to have e-commerce orders in a Security dashboards, nevertheless, it is also possible to change the Data View in the Timeline, so we are not limited to render alerts only.

kertal commented:
> Hi @kertal Yes you are right, the scenario you described is the expected behavior. However, these actions will only be visible inside Security, where the dashboards are expected to contain Security-related data. The "add to timeline" action is meant to make it easy for Security analysts to pivot to Timeline investigations on an specific value, probably from the alerts index. It does not make much sense to have e-commerce orders in a Security dashboards, nevertheless, it is also possible to change the Data View in the Timeline, so we are not limited to render alerts only.

@semd  Thx, this makes sense. Quick follow up, what's the quickest way to generate test data for the security related data?
@lukasolson could you also have a look, this PR has some overlapping with your work to extend the doc viewer , many thx

<img width="1740" alt="Bildschirmfoto 2023-05-15 um 15 24 04" src="https://github.com/elastic/kibana/assets/463851/4f349e8e-ed96-424f-b9f0-c33f48539c4c">

semd commented:
@kertal What I usually do to generate some alerts data locally is executing the test generator with:
```
cd x-pack/plugins/security_solution
yarn test:generate -s resolver --gen 3 --pr 100 --pt 100
```
And then create a new rule in the UI (_/app/security/rules/create_) that takes everything `*` using the default index patterns, to generate the alerts.

semd committed:
relocate cellActions context provider

semd commented on src/plugins/discover/public/application/services_provider.tsx:
```diff
@@ -0,0 +1,26 @@
+  const { getTriggerCompatibleActions } = services.uiActions;
+  return (
+    <KibanaContextProvider services={services}>
+      <CellActionsProvider getTriggerCompatibleActions={getTriggerCompatibleActions}>
```
@jughosta Yes that makes sense, I moved the `CellActionsProvider` inside the `DiscoverDocuments` component wrapping only the `DiscoverGrid`. This new services_provider is no longer needed, removed.

semd replied:
@jughosta I had to use the provider inside `src/plugins/discover/public/application/context/context_app_content.tsx` as well

jughosta replied:
There is also an option to defined a default value for the context. Example:

```typescript
const DEFAULT_ACTIONS = []
const CellActionsContext = createContext<{ getActions: GetActions }>({
  // set a default handler for cases when provider is not used
  getActions: () => Promise.resolve(DEFAULT_ACTIONS),
});

export const CellActionsProvider: FC<CellActionsProviderProps> = ({
  children,
  getTriggerCompatibleActions,
}) => {
  const value: { getActions: GetActions } = useMemo(
    () => ({
      getActions: (context) =>
        getTriggerCompatibleActions(context.trigger.id, context).then((actions) =>
          orderBy(['order', 'id'], ['asc', 'asc'], actions)
        ) as Promise<CellAction[]>,
    }),

    [getTriggerCompatibleActions]
  );

  // make sure that provider's value does not change
  return <CellActionsContext.Provider value={value}>{children}</CellActionsContext.Provider>;
};

// don't trigger an exception
export const useCellActionsContext = () => useContext(CellActionsContext);
```

semd replied:
That's smart. 
But doing this would hide incorrect uses of the `CellActions` framework, making it harder to debug as well. The exception is intended, to make sure `CellActions` are only used inside its Provider, so we notice if missing during the development stage. We already had this discussion within the team and we decided to be strict.
(That's also why I put it at the top level from the beginning)

kertal commented:
> @kertal What I usually do to generate some alerts data locally is executing the test generator with:
> 
> ```
> cd x-pack/plugins/security_solution
> yarn test:generate -s resolver --gen 3 --pr 100 --pt 100
> ```
> 
> And then create a new rule in the UI (_/app/security/rules/create_) that takes everything `*` using the default index patterns, to generate the alerts.

@semd thx, now it makes much more sense for me, and looking at it, I think we should align styling and wording, here's a screen to compare

<img width="559" alt="Bildschirmfoto 2023-05-16 um 09 24 01" src="https://github.com/elastic/kibana/assets/463851/bc9d16f9-8631-4aea-9afb-bb3f38569202">

semd committed:
add cellActions provider to context_app

semd committed:
Merge remote-tracking branch 'upstream/main' into poc_discover_custom_cell_actions

semd commented:
> I think we should align styling and wording

Hi @kertal, this is the wording and styling that we use everywhere in Security, the inconsistency was in the saved searches embeddable displayed in the "Security dashboards" page, they were showing the Discover-styled cell actions. This PR solves this inconsistency. Cell actions will be consistent(*) within Security, and they will also stay consistent within Discover.
I agree that there are differences between Security and Discover apps, but I think that's another topic.

(*) The document flyout opened by saved search embeddable still renders Discover-styled actions, we can solve that in a separate PR if needed.

kertal commented:
> > I think we should align styling and wording
> 
> Hi @kertal, this is the wording and styling that we use everywhere in Security, the inconsistency was in the saved searches embeddable displayed in the "Security dashboards" page, they were showing the Discover-styled cell actions. This PR solves this inconsistency. Cell actions will be consistent(*) within Security, and they will also stay consistent within Discover. I agree that there are differences between Security and Discover apps, but I think that's another topic.
> 
> (*) The document flyout opened by saved search embeddable still renders Discover-styled actions, we can solve that in a separate PR if needed.

Thx, yes, make sense, if this is Security wording, it should stay this way. Document flyout can be solved in a follow up, I think this makes sense 👍

semd committed:
use original DataGrid in test

angorayc commented:
Tested locally ✅ 
https://github.com/elastic/kibana/assets/6295984/4b7c388a-15ed-46a9-b1cf-501cb1f954bf

Some minor observations we could consider deal with it in the future:
Copy value and Copy to clipboard might look similar to users but they have different behaviour.

- Copy value - one of the default action from Kibana Dashboard. It copys values only (no double quote). e.g.: MacBook-Pro.local
- Copy to clipboard - Security Solution custom action, it copys `field`: `value`. e.g.: host.name: "MacBook-Pro.local" 

The actions in Kibana dashboard is a bit different between legend actions and cell actions:
- Legend Actions do not contain `Copy value`
<img width="2557" alt="Screenshot 2023-05-17 at 10 37 57" src="https://github.com/elastic/kibana/assets/6295984/0824a175-7189-4392-ab67-b5c71b7def3d">
<img width="2559" alt="Screenshot 2023-05-17 at 10 38 06" src="https://github.com/elastic/kibana/assets/6295984/c736bb02-eb0b-4f24-9999-7c801efa0d61">


The actions in Security Solution dashboard looks aligned: We have `Copy to clipboard` for both legend actions and cell actions.
<img width="2542" alt="Screenshot 2023-05-17 at 10 39 15" src="https://github.com/elastic/kibana/assets/6295984/219be21a-41d8-416d-871a-b2255d37f6d1">
<img width="2485" alt="Screenshot 2023-05-17 at 10 39 29" src="https://github.com/elastic/kibana/assets/6295984/c8432575-4413-4886-8313-210685e86f32">

semd committed:
getCellValue function prop

michaelolo24 commented on x-pack/plugins/security_solution/public/actions/register.ts:
```diff
@@ -44,12 +58,47 @@ const registerLensActions = (store: SecurityAppStore, services: StartServices) =
   uiActions.addTriggerAction(CELL_VALUE_TRIGGER, copyToClipboardAction);
 };
 
+const registerDiscoverCellActions = (store: SecurityAppStore, services: StartServices) => {
```
👌🏾

michaelolo24 reviewed (approved):
Thank you for this work, it's gonna be really helpful for us in the future!

jughosta commented on src/plugins/discover/public/components/discover_grid/discover_grid.tsx:
```diff
@@ -416,14 +426,49 @@ export const DiscoverGrid = ({
+    () =>
+      cellActionsTriggerId
+        ? visibleColumns.map((columnName) => {
+            const field = dataView.getFieldByName(columnName)?.spec;
```
Discover works in 2 modes: a regular (simply for data views) one and for text-based search (for example SQL). And users can enable SQL via `discover:enableSql` Advanced Setting and then also save a search on Discover as a saved search which can be added to a dashboard. 

Retrieving a field from a data view will work for the regular mode but unfortunately not for text based mode as columns can be named by user via custom variables. Information about columns is returned by Elasticsearch in a different format and it can be accessed in `discover_documents.tsx` under `documentState.textBasedQueryColumns` if necessary. `isPlainRecord` variable indicates that the current mode is text-based.

Not sure though if it makes sense in Security Solutions to enable special cell actions for text-based searches. If not, then we could return `undefined` fields here if  `isPlainRecord` like the following:
```
() => cellActionsTriggerId && !isPlainRecord ? visibleColumns.map... : undefined
```

jughosta reviewed (commented):
Thanks a lot for the performance optimizations! Left more suggestions in comments. 

Nit: would be great to add some padding to the Security popover if possible:
- discover page
<img width="348" alt="Screenshot 2023-05-22 at 15 58 27" src="https://github.com/elastic/kibana/assets/1415710/9627714e-0957-4381-b83d-5605f7203e0d">

- vs security page
<img width="269" alt="Screenshot 2023-05-22 at 15 58 48" src="https://github.com/elastic/kibana/assets/1415710/7d50281b-6ede-4198-a5b0-024420dacc86">

lukasolson reviewed (commented):
This is awesome! Thanks for the example of how to integrate cell actions in Discover, as this is definitely something we had planned on doing in the near future anyway.

Quick question: Does it make sense to register these inside of the `setup` phase instead of the `start` phase? It doesn't make a big difference right now since they are only used in the security plugin, but if I, for example, want to register a cell action that shows up in other plugins, registering in the `start` phase wouldn't actually make it show up, correct?

semd commented on packages/kbn-cell-actions/src/types.ts:
```diff
@@ -20,6 +20,7 @@ export interface CellActionsProviderProps {
   getTriggerCompatibleActions: UiActionsService['getTriggerCompatibleActions'];
 }
 
+export type CellActionValue = string | string[] | null | undefined;
```
This can be a problem, all cell actions expect this type currently. In Security turn them into strings, and everything works, are you okay with that? Otherwise, we'd have to refactor all the cell actions registered.

jughosta replied:
Sorry, I am not familiar with how/why cell action logic is tied to string type only. Could you please link to that part of code?

There is this type casting in the PR which looks problematic as values coming from users data can be not only strings.
```
return displayedRows[pageRowIndex].flattened[fieldName] as CellActionValue;
```

semd replied:
All the actions in Security:
https://github.com/elastic/kibana/tree/bb495c2d5088456208ca26980d7c4c122f2743d9/x-pack/plugins/security_solution/public/actions

and the generic ones in the package as well:
https://github.com/elastic/kibana/tree/bb495c2d5088456208ca26980d7c4c122f2743d9/packages/kbn-cell-actions/src/actions

expect the `CellActionValue` type in the value parameter for the `isCompatible` and `execute` functions, if we add the number and object types to it we would have to adapt all those actions. That would be way out of the scope of this task.

We can either block this PR until we add support for those types to the `CellActionValue` (not sure it is going to happen anytime soon) or we can try to find a workaround:

In Security we convert numbers to strings and everything works, it should not be a problem. But we do not have objects, we could try to stringify, but I think it would be better to just hide the actions for object values somehow. and use the default discover cell actions, that support objects.

What do you think?

jughosta replied:
> In Security we convert numbers to strings and everything works, it should not be a problem. But we do not have objects, we could try to stringify, but I think it would be better to just hide the actions for object values somehow. and use the default discover cell actions, that support objects.

Sounds good to me if it fits Security context.

semd commented:
> This is awesome! Thanks for the example of how to integrate cell actions in Discover, as this is definitely something we had planned on doing in the near future anyway.
> 
> Quick question: Does it make sense to register these inside of the `setup` phase instead of the `start` phase? It doesn't make a big difference right now since they are only used in the security plugin, but if I, for example, want to register a cell action that shows up in other plugins, registering in the `start` phase wouldn't actually make it show up, correct?

Thanks @lukasolson, 
Security actions need to use resources that are only available after mounting the app, such as the history object or the redux store. So you are right that if Security is never accessed those actions will never be registered, but that's okay for us since those actions are specific for Security, and they are not supposed to be used from outside the app. 

However, we are not registering the actions in the plugin `start` directly, they are registered in the application `mount`, so only when Security is not mounted (accessed at some point) they are registered, if we were registering in the `start` lifecycle directly, which is called for all enabled plugins when Kibana starts, the Security app won't need to be accessed in order to have the actions registered.
 
That said, other plugins will have those Security-specific actions available as long as their components are rendered inside Security Solution. 
For other solutions (e.g. Observability) I don't think they should use actions owned and registered by another solution, this would create (hidden) cross-solution dependencies, they would have no control over the implementation, and we would have to be very careful not to break anything from other solutions when doing changes. 

If we were to share the same generic actions and triggers in multiple solutions, a better option would to create and register them from a generic place, like in the Discover plugin itself if the actions were going to be used by Discover embeddable. That would be fine for very generic use cases, however, other solutions would still have any control over those actions and triggers. We won't be able to add a custom cell action to the trigger, for instance.

I think the best way for solutions to use generic actions and have full control is to use the generic action factories we already provide in the cell actions package:

https://github.com/elastic/kibana/blob/667ec95d301b28e49400d0fb05fd8e887d2dce7b/packages/kbn-cell-actions/src/index.ts#L31-L37

We have the 3 basic ones ("Filter In", "Filter Out" and "Copy to clipboard") but we can add more. 
They can be imported and registered and they will work out of the box, they can also be extended. This way we delegate the control over those generic actions to the solutions, they can register a combination of generic and specific actions to a trigger, and we don't introduce any dependency on other plugins.

Does it make sense?

semd committed:
merge main

semd committed:
address review comments

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

semd committed:
merge main

semd committed:
Merge branch 'poc_discover_custom_cell_actions' of https://github.com/semd/kibana into poc_discover_custom_cell_actions

semd commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into poc_discover_custom_cell_actions

semd commented:
> would be great to add some padding to the Security popover if possible:
@jughosta The popover text hasn't changed, it is not a Security thing, it is the way the Saved Search embeddable renders the cell actions popover.

jughosta commented:
> > would be great to add some padding to the Security popover if possible
> 
> @jughosta The popover text hasn't changed, it is not a Security thing, it is the way the Saved Search embeddable renders the cell actions popover.

Looks like it's because this style override: https://github.com/semd/kibana/blob/d8fe39c18d9df7acb00a82515f5a41a6e4111c59/x-pack/plugins/security_solution/public/common/components/page/index.tsx#L62

`EuiDataGrid ` now supports `setCellPopoverProps({ panelClassName: <...> })` for customization which could be used instead of overriding global styles.

semd commented:
> Looks like it's because of this style override: https://github.com/semd/kibana/blob/d8fe39c18d9df7acb00a82515f5a41a6e4111c59/x-pack/plugins/security_solution/public/common/components/page/index.tsx#L62
> 
> `EuiDataGrid` now supports `setCellPopoverProps({ panelClassName: <...> })` for customization which could be used instead of overriding global styles.

Oh yes, good catch. Now I remember we have these overrides here, they are so annoying. I am going to create a separate task to remove this global style and use the className. But I don't think it is a good idea to do it in this PR, this style affects all the cellActions in the entire Security application, we have a lot of them, and this style might be there for a reason.

jughosta commented on src/plugins/discover/public/components/discover_grid/discover_grid.tsx:
```diff
@@ -432,14 +442,65 @@ export const DiscoverGrid = ({
+
+  const getCellValue = useCallback<UseDataGridColumnsCellActionsProps['getCellValue']>(
+    (fieldName, rowIndex): CellActionValue => {
+      const value = displayedRows[rowIndex % displayedRows.length].flattened[fieldName] as
+        | CellActionValue
+        | number
+        | number[];
+      if (value == null) {
+        return value;
+      }
+      if (Array.isArray(value)) {
+        return value.map((val) => val.toString());
+      }
+      return value.toString();
```
This value format is what security cell actions expect, right? What if other consumers of cell actions integration would like to register their actions and have access to the original field values? I think the transformation should not be a part of Discover code.

semd replied:
Actually, this is the format that the CellActions framework expects, we tried to make it as generic as possible, but I admit we couldn't help but have some Security Solution bias.
Nevertheless, this is the same problem we discussed [here](https://github.com/elastic/kibana/pull/157201#discussion_r1200457347)
If other consumers of cell actions integration would like to register their actions and receive `number | number[]` value type we would have to refactor all the actions registered currently.
(I would not consider adding the `_source` type because is too specific to Discover.)

semd replied:
actions in Security:
https://github.com/elastic/kibana/tree/bb495c2d5088456208ca26980d7c4c122f2743d9/x-pack/plugins/security_solution/public/actions

generic actions in the package:
https://github.com/elastic/kibana/tree/bb495c2d5088456208ca26980d7c4c122f2743d9/packages/kbn-cell-actions/src/actions

I have been testing numeric values with all the actions we have and everything works well, elastic does not care about quoting numbers.

![string_numbers](https://github.com/elastic/kibana/assets/17747913/ac157b3e-292b-4733-96b9-fa995d83fc79)
![string_numbers_2](https://github.com/elastic/kibana/assets/17747913/0321028a-4d05-4736-9aca-ed355993c7f2)

I don't think we should refactor everything now, we haven't got any problems so far using this approach. If we find a use case for number type is needed we'll add support for it, I don't think we should worry about it for now.

semd replied:
@jughosta Ironically enough, while testing it I found a bug with an action.
I created a separate issue here https://github.com/elastic/kibana/issues/159298 to do the refactor and add number support.
If you want we can merge this PR as it is, and I will add a checkpoint in the issue I just created to come to Discover and remove the string casting I added.
Does it sound good?

semd committed:
Merge remote-tracking branch 'upstream/main' into poc_discover_custom_cell_actions

semd committed:
account for array of null | undefined

semd commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into poc_discover_custom_cell_actions

davismcphee reviewed (commented):
This looks great overall, and I like the idea of building on top of UI actions to standardize cell actions, but I share the same concerns as @jughosta.

Cell actions were introduced to us as a general purpose framework for handling actions in grid cells, which is an idea I really like, but in their current state they don't seem to be usable outside of the Security use case -- or at least we can't make use of them in Discover, which is important if we're introducing them into Discover.

Part of our willingness to introduce cell actions into Discover was with the understanding that they'd provide a benefit to us and we'd be able to migrate Discover's default actions to cell actions, as well as using them for our Unified Doc Viewer and Unified Data Table components. But with the current lack of support for all value types, it feels more like a system tacked on to Discover just to fulfill the Security use case.

>What do you mean by "raw" data? CellActions is not supposed to accept `unknown` values, we need to define a specific type. If you block this PR for this it will probably stay blocked forever.

This is concerning since in order to support all of the types that we might encounter, cell actions would effectively need to support all types that could be returned from ES and leave type checks and conversions to the actions themselves.

>I don't think we should expect the "perfect" solution from the start, we can do a "good enough" approach for now, which is still better than nothing (only Security uses this and it is okay for us), and then we can iterate it and improve it.

I agree that we don't need a perfect solution from the start, but we do need a clear path forward to making this feature useful outside of the context of Security since introducing it into Discover opens it up to broader use.

>Looks like the only problem is the object type, but no action we currently have will work with objects, we could add it in the type but then we'll have to exclude it everywhere via compatibility checks inside all actions.

This sounds like a more robust way to handle it over forcing consumers to naively convert all values to `string` -- although personally I would just use `Serializable` from `packages/kbn-utility-types/src/serializable/index.ts` instead of expanding `CellActionValue`. An alternative might be to introduce something like a `convertValue` function directly onto the cell action definition for actions that need it, although that's just off the top of my head and I haven't looked into how that would work type wise.

>I would rather not do that if there is no specific use case for object type. If we ever have an action that would work with objects can do that, but there's no point in accepting something that will require a ton of extra logic just to ignore it, and which is not going to be used anywhere.
So I think it is better to just omit the object type and let the default Discover cell actions render instead. I already did that with the `_source` type, is there any other case in which we could receive an object value? Can we detect it via dataView `spec` or something?

An example of where `object` would be needed is for `flattened` field types. In Discover we have a `Copy value` action that supports `flattened` field types, but with cell actions the copied value would get converted to `[object Object]`:
<img width="753" alt="flattened_action" src="https://github.com/elastic/kibana/assets/25592674/9ff7ae99-0579-484f-b823-79d826a35da8">
<img width="433" alt="flattened_console" src="https://github.com/elastic/kibana/assets/25592674/07754776-92e4-4835-aef6-de40f6ce17ac">

Ultimately I won't block merging this PR over the type issue if we can agree that cell actions will need to support everything covered by `Serializable` as a followup, otherwise I worry this framework just isn't useful to have in Discover for anyone outside of Security -- and I would personally love to be able to use it since I think it's a great framework that just currently has a shortcoming.

On a different note, I did some local testing and have one potential bug to report as well as a question:
- I can no longer add a saved search with a column that has the `flattened` field type to a Dashboard. Instead of loading the saved search, I get an empty panel with some console errors:
  - In the `main` branch:
    ![flattened_working](https://github.com/elastic/kibana/assets/25592674/b23dd863-1a1f-49a4-adf0-cb548967139f)
  - In this branch:
    ![flattened_broken](https://github.com/elastic/kibana/assets/25592674/62a10ca7-9bef-4e83-9e34-28c2ac59f631)
- As far as I can tell the "Filter In" and "Filter Out" actions weren't doing anything when I clicked them. I'm assuming this might be related to me using non-Security data for testing, but is it expected that nothing happens when clicking these actions?
  ![filter_buttons](https://github.com/elastic/kibana/assets/25592674/92e3c62b-09c3-4d1b-abf5-7ab0ad3f89bf)

semd committed:
Merge remote-tracking branch 'upstream/main' into poc_discover_custom_cell_actions

semd commented:
@davismcphee Thank you so much for reviewing and testing the PR. I addressed the nits. 
About the bugs you mentioned:

- I have tested _number_ fields and they work correctly, I am not sure what value the "Filter" actions are receiving for this "response" column. Could you let us know how to test it?

- I excluded the `flattened` field type the same way as the `_source` field for now, Discover actions will be displayed for these columns. But I haven't been able to test it either, could you let us know how to ingest data locally to reproduce it?

Regarding the support for `Serializable`. I think we are missing a couple of important considerations here:

1- CellActions use the field properties to determine the compatibility (`name`, `type`, ...), we can not use the `value` because the EuiDataGrid needs the columnCellActions to be defined by the whole column, not for each cell, so we can not do `typeof value` to know if the action is compatible. When we have the actual value of the cell, the action is already rendered.

2- (The most important) CellActions framework is built to let the user pivot, by performing different "query" actions: 

- "Filter in/out" add _queries_ to the filterManager, 
- "AddToTimeline" adds a dataProvider _query_ to the Timeline
- "CopyToClipboard" copies the value in a _query_ format, so it can be pasted to any search bar
- "ShowTopN" _queries_ the different top values of the field and displays a chart.

So, in order for actions to understand what they are expected to _query_, we need to use a structured value type. As an example, all actions know that if they receive an array value it means that the cell is multi-valuated, so they need to apply the `AND` conditional between all the values inside the array. We should not send an arbitrary object and array values to the actions and expect they will be treated as plain values, that's not the way it works. 
The `Serializable` type allows recursively nested objects and arrays, we can not support that as we intended the CellActions. 

If we decide to support those "arbitrary" object and array values in some of the actions, I think we should just stringify them and, once inside the action, check the `field.type` to process them differently if needed. We'd also need to add a compatibility check by `field.type` in the actions that do not support those values.

semd committed:
Merge branch 'poc_discover_custom_cell_actions' of https://github.com/semd/kibana into poc_discover_custom_cell_actions

semd committed:
PR suggestions

jughosta commented:
[There are](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-types.html) more complex field types than simple ones. Let's start then with rather whitelisting `field.type`s for which we know that it will work predictably and correctly. It could be `string`(`keyword`, `text`), `number` and most probably `ip` and `date` too. It's not necessary to transform their values. Filter manager can work with any value by using `generateFilters`: https://github.com/elastic/kibana/blob/ba6f36757ee39321e29d0093f8412223bfad2dad/src/plugins/data/public/query/filter_manager/lib/generate_filters.ts#L81 Example of how it's implemented in Discover https://github.com/elastic/kibana/blob/633444e6158d87983f0314c2a214af0e791d48d9/src/plugins/discover/public/application/main/components/layout/discover_layout.tsx#L142 It would be great if cell actions start using `generateFilters` util for FilterIn/Out too. 
Additionally to this, do we have checks for `field.filterable` before showing filter actions?

semd commented:
@jughosta Whitelisting types sound like a good approach to me. The ones you mentioned are all supported, `version` and other text types can also be included.

We are currently working on making the `field` properly more generic in the CellActions [here](https://github.com/elastic/kibana/pull/157834), the PR changes the field type to `FieldSpec`, but maybe we should consider changing that to use the entire `DataViewField`, so we would be able to use getters such as `filterable`, `displayName`... 
cc @machadoum

davismcphee commented:
Thanks for taking care of the nits!

> * I have tested _number_ fields and they work correctly, I am not sure what value the "Filter" actions are receiving for this "response" column. Could you let us know how to test it?

No problem, I tested by installing "Sample web logs" sample data, creating a new dashboard with only a "[Logs] Visits" saved search panel, then navigated to `Security > Dashboards > CUSTOM` and selected my created dashboard. I then tried clicking the filter cell action buttons but it didn't seem like anything was happening. The `response` field is a `text` field and looked like this within the action `execute` function (although I tried on multiple fields and didn't notice a difference):
<img width="493" alt="filter" src="https://github.com/elastic/kibana/assets/25592674/5da99234-cd96-4c9f-af18-7063574ddcde">

> * I excluded the `flattened` field type the same way as the `_source` field for now, Discover actions will be displayed for these columns. But I haven't been able to test it either, could you let us know how to ingest data locally to reproduce it?

Sure thing, this is what I did for this one:
- Go to the Kibana console and execute the first example code block on the [flattened field docs page](https://www.elastic.co/guide/en/elasticsearch/reference/current/flattened.html) to ingest the data.
- Create a data view for the `bug_reports` index in Discover, and then create a saved search pointing at the data view.
- Navigate to `Security > Dashboards > CUSTOM` and click "Create Dashboard".
- Try to add the previously created saved search and it should fail to load.

> 1- CellActions use the field properties (`name`, `type`, ...) in the compatibility check, but we can not use the `value` to determine compatibility. That's because the EuiDataGrid needs the `columnCellActions` to be defined for the whole column, not for each cell, so we can not do `typeof value` to know if the action is compatible. When we have the actual value of the cell is when the action's component is rendering, but it is already in the `columnCellActions` array, so we can not return `null` because it affects the `visibleCellActions` count anyway.

Sorry for the confusion here, I wasn't meaning to say that all actions would need to support all types or that access to values in compatibility checks are necessary. What I meant was that the cell actions framework as a whole should have the capability to support actions for any value that could be returned from ES. That way it would be left up to each action's compatibility check to exclude field types it doesn't support by checking the `field.type` property. I don't see it as necessary for the entire framework to exclude most value types, it feels more like an artificial limitation to me.

> 2- (The most important) CellActions framework is built to let the user do actions with the cell value, and majority of those actions imply doing some kind of _query_ with the value:
> 
> * "Filter in/out" add _queries_ to the filterManager,
> * "AddToTimeline" adds a dataProvider _query_ to the Timeline
> * "CopyToClipboard" copies the value in a _query_ format, so it can be pasted to any search bar
> * "ShowTopN" _queries_ the different top values of the field and displays a chart.

This is the case for the actions the Security team currently uses, but this isn't true for all potential consumers of cell actions. For example our [`CopyBtn`](https://github.com/elastic/kibana/blob/f1db0cd851d387091a836d4150c115648131a9e2/src/plugins/discover/public/components/discover_grid/discover_grid_cell_actions.tsx#L89-L118) action doesn't copy values to a query format, and instead accepts `unknown` values and stringifies them differently based on the value type before copying them to the clipboard. The way the cell actions framework currently works would prevent us from migrating to it for our own actions even though it's integrated into Discover.

> So, in order for actions to understand what they are expected to _query_, we need to use a structured value type. As an example, all actions know that if they receive an array value it means that the cell is multi-valuated, so they need to apply the `AND` conditional to all the values inside the array. We should not send an arbitrary object and array values to the actions and expect they will be treated as plain values, that's not the way CellActions is intended. The `Serializable` type allows recursively nested objects and arrays, which we can not support because they do not represent multiple values, it is just one value in a "serializable" format. We need a way to discriminate between both scenarios.
>
> If we decide to support those arbitrary object and array values in some of the actions, I think we should just stringify them and, once inside the action, check the `field.type` to process them differently if needed. We'd also need to add a compatibility check by `field.type` in the actions that do not support those values.

They aren't arbitrary values though, they are the raw values returned by ES. You could still rely on your array checks, and the [`generateFilters`](https://github.com/elastic/kibana/blob/f1db0cd851d387091a836d4150c115648131a9e2/src/plugins/data/public/query/filter_manager/lib/generate_filters.ts#L81-L129) function Julia linked to also does that. The important part is that actions would use `field.type` to determine compatibility and how to handle values, rather than relying on all values being preprocessed before being passed to the framework.

I don't see the benefit of stringifying the values before passing them to the framework since you would need to do type checks anyway to determine whether to stringify with `.toString()` for primitives or `JSON.stringify()` for complex types. Then you'd also have to do the opposite from within the action to get the unstringified value. Avoiding any preprocessing would allow us to use the same logic you propose by checking `field.type` for compatibility checks, and to determine which value types to expect within the actions.

> [There are](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-types.html) more complex field types than simple ones. Let's start then with rather whitelisting `field.type`s for which we know that it will work predictably and correctly. It could be `string`(`keyword`, `text`), `number` and most probably `ip` and `date` too. It's not necessary to transform their values. Filter manager can work with any value by using `generateFilters`:

@jughosta I'm curious were you thinking of allowlisting field types as part of the action compatibility checks, or within `discover_grid` before we pass them to the cell actions framework?

jughosta commented:
@davismcphee 

> @jughosta I'm curious were you thinking of allowlisting field types as part of the action compatibility checks, or within discover_grid before we pass them to the cell actions framework?

Depends what is feasible at the moment to prevent stringifying the values and to avoid unexpected errors. Ideally it should be a part of security action compatibility checks as other future actions might want to support any fields values.

semd commented:
@davismcphee @jughosta 
Okay, that seems reasonable, I agree that this design will make the CellActions more generic. It means a lot of work because it implies updating all the registered actions, but we can do it in a separate PR if you agree to stay in an intermediate state while we do the changes.
We could follow this roadmap:

Stage 1 (this PR):
- Fix any remaining bugs.
- Whitelist the `field.type`s supported by CellActions in the `DiscoverGrid` component for now. The default Discover actions will be rendered for the rest of the types.

Stage 2 (follow-up PR):
- Change the CellActions type to `value: Serializable`.
- Whitelist the `field.type` in the `isCompatible` of every action.
- Add the proper validations and value castings in the `execute` of every action.
- Remove the whitelist from `DiscoverGrid` and pass the value directly to CellActions.

Based on the roadmap for the Security serverless integration with Discover, it would be highly desirable to complete stage 1 by the end of June. We are committed to starting stage 2 upon the merging of stage 1.

I do not have enough bandwidth to absorb this task, but @machadoum is going to take over this PR and the following ones, he is the main author of the CellActions package ⭐ 

What do you think, do we have a plan?

davismcphee commented:
@semd Thanks for discussing this with us, and that plan sounds great to me. I appreciate that Security needs to iterate on the Discover work and this PR is blocking progress, so it definitely makes sense to separate the followup work for later. In that case this PR should be good to go on my end once we've added the temporary allowlist to `DiscoverGrid` and addressed any outstanding bugs (I'm thinking the allowlist will probably solve the `flattened` one I ran into anyway). Thanks again for the work on this and I'm looking forward to having cell actions in Discover soon!

machadoum committed:
Create an allowlist for supported CellActions field types displayed on DiscoverGrid

machadoum committed:
Merge remote-tracking branch 'elastic/main' into poc_discover_custom_cell_actions

	src/plugins/discover/tsconfig.json

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

machadoum committed:
Fix unit test

machadoum commented:
Hey team!  As Sergi mentioned, I am taking it over from here.

First of all, thank you for helping us improve CellActions ❤️ 

I pushed code with KBN types allowlist. We are only going to display `CellActions` for the types: `date`, `ip`, `string`, and `number`. Could you please take a look https://github.com/elastic/kibana/pull/157201/commits/5e4ac3108efd0a78604c6f4cfae05f320fca0d50?

@davismcphee I tried to reproduce the response field bug step-by-step, but it works for me. I suspect that a previous change fixed it. Could you try to reproduce it one more time with the current changes?


https://github.com/elastic/kibana/assets/1490444/ce4069d1-4a99-4ce6-b51f-39782d284cd9

jughosta reviewed (commented):
Hi @machadoum,
Thanks for continuing the work and planning on extending cell actions package with more types!

jughosta reviewed (approved):
Unblocking the PR 👍

davismcphee reviewed (changes requested):
@machadoum Thanks for the updates, the allowlist change looks good to me! Overall this PR seems just about ready to merge, except I'm still encountering the issue with adding saved searches containing a `flattened` field to a Dashboard.

Here's a quick video of what I'm doing from start to finish to reproduce the issue: https://github.com/elastic/kibana/assets/25592674/5904ba82-f015-4263-98f3-e02da83ff696

>@davismcphee I tried to reproduce the response field bug step-by-step, but it works for me. I suspect that a previous change fixed it. Could you try to reproduce it one more time with the current changes?

I'm now thinking this might just be a misunderstanding on my part and maybe this is expected. I notice in your video there's a search bar above the dashboard, but I'm not seeing one when I view my dashboard, which probably explains why clicking the filter actions don't do anything for me:
<img width="1823" alt="dashboard" src="https://github.com/elastic/kibana/assets/25592674/d5eac5db-8f22-4f4d-9a10-16a2d786fc10">

This one is less of a concern for me anyway since it doesn't have an impact outside of Security, so I'd be happy to merge with the current behaviour if it's working as expected for your team. The only thing left I'd like to make sure we figure out before I approve is the blank saved search panel since it impacts any dashboard configured this way.

machadoum commented on src/plugins/discover/public/components/discover_grid/discover_grid_cell_actions.tsx:
```diff
@@ -120,3 +121,12 @@ export const CopyBtn = ({ Component, rowIndex, columnId }: EuiDataGridColumnCell
   return [...(onFilter && field.filterable ? [FilterInBtn, FilterOutBtn] : []), CopyBtn];
 }
+
+// Converts the cell action value to the type expected by CellActions component
+export const convertCellActionValue = (rawValue: unknown): CellActionValue => {
+  const value = rawValue as CellActionValue | number | number[];
+  if (Array.isArray(value)) {
+    return value.map((val) => (val != null ? val.toString() : val));
+  }
+  return value != null ? value.toString() : value;
+};
```
Yes, I already have a draft PR for adding numbers. I will delete it soon.

machadoum committed:
Fix blank disciver grid when data view has no timestamp

machadoum committed:
Please code reviewers

machadoum commented:
Hey @davismcphee!
Thank you for the video it was very helpful. I managed to reproduce the bug. It happened when the data view had no times field. I noticed that before this PR there was an extra validation for it, so I added it back [here](https://github.com/elastic/kibana/pull/157201/commits/23fec203220ab04826aa11b4ecf8987318319367#diff-2ff8a511b1d6875663f83e10aff58f096d4887c195ee5d501753e7ac30d560b3R244).  Good catch!

davismcphee reviewed (approved):
I tested with the latest changes and the blank panel issue is fixed! Also thanks for adding those tests to help prevent it in the future. I'm glad you were able to get to the bottom of the missing search bar, I'll leave it to you to decide if that's something you need to address in this PR or not. Thanks for all the work on this, LGTM 👍

machadoum committed:
Merge remote-tracking branch 'elastic/main' into poc_discover_custom_cell_actions

machadoum committed:
Fix merge conflicts

machadoum commented:
@elasticmachine merge upstream

machadoum committed:
Merge remote-tracking branch 'elastic/main' into poc_discover_custom_cell_actions
